src/pulsar_producer.erl

```diff
@@ -61,106 +99,227 @@ callback_mode() -> [state_functions].
                 batch_size = 0,
                 requests = #{},
                 last_bin = <<>>}).
+-type state() :: #state{
+                    partitiontopic :: binary(),
+                    broker_server :: binary(),
+                    sock :: undefined | port(),
+                    request_id :: integer(),
+                    producer_id :: integer(),
+                    sequence_id :: sequence_id(),
+                    producer_name :: atom(),
+                    opts :: map(),
+                    callback :: undefined | mfa() | fun((map()) -> ok),
+                    batch_size :: non_neg_integer(),
+                    requests :: #{sequence_id() => [{replayq:ack_ref(),
+                                                     [gen_statem:from()],
+                                                     [{timestamp(), [pulsar:message()]}]}]},
+                    last_bin :: binary()
+                   }.
+
+callback_mode() -> [state_functions, state_enter].
```

Comment:
When does this callback_mode change take effect?

Reply:
According to the [docs](https://www.erlang.org/doc/design_principles/statem.html#callback-modes), seems possible.  In this other [doc](https://www.erlang.org/doc/man/gen_statem.html#Module:code_change-4), they say that it's required to use `{advanced, _}` in the appup, otherwise the change in callback mode won't be honored.

> Also note when upgrading a gen_statem, this function and hence the `Change = {advanced,Extra}` parameter in the [appup](https://www.erlang.org/doc/man/appup.html) file is not only needed to update the internal state or to act on the Extra argument. It is also needed if an upgrade or downgrade should change [callback mode](https://www.erlang.org/doc/man/gen_statem.html#type-callback_mode), or else the callback mode after the code change will not be honoured, most probably causing a server crash.